src/phoenix/session/client.py

```diff
@@ -249,6 +264,142 @@ def log_traces(self, trace_dataset: TraceDataset, project_name: Optional[str] =
                 },
             ).raise_for_status()
 
+    def upload_dataset_csv(
+        self,
+        file: Union[str, Path],
+        input_keys: Iterable[str] = (),
+        output_keys: Iterable[str] = (),
+        metadata_keys: Iterable[str] = (),
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+        action: Literal["create", "append"] = "create",
+    ) -> None:
+        """
+        Upload CSV file as dataset to the Phoenix server.
+
+        Args:
+            file (str | Path): Location of the csv file.
+            input_keys (Iterable[str]): List of column names used as input keys.
+                input_keys, output_keys, metadata_keys must be disjoint, and must
+                exist in CSV column headers.
+            output_keys (Iterable[str]): List of column names used as output keys.
+                input_keys, output_keys, metadata_keys must be disjoint, and must
+                exist in CSV column headers.
+            metadata_keys (Iterable[str]): List of column names used as metadata keys.
+                input_keys, output_keys, metadata_keys must be disjoint, and must
+                exist in CSV column headers.
+            name: (Optional[str]): Name of the dataset. Required if action=append.
+            description: (Optional[str]): Description of the dataset.
+            action: (Literal["create", "append"): Create new dataset or append to an
+                existing dataset. If action=append, dataset name is required.
+
+        Returns:
+            None
+        """
+        if action not in ("create", "append"):
+            raise ValueError(f"Unknown action: {action}")
+        if action == "append" and not name:
+            raise ValueError(f"Dataset name must not be empty for {action=}")
+        keys = DatasetKeys(frozenset(input_keys), frozenset(output_keys), frozenset(metadata_keys))
+        path = Path(file).resolve()
+        if not path.is_file():
+            raise ValueError(f"File does not exist: {path}")
+        with open(path, "r") as f:
+            for row in csv.reader(f):
+                column_headers = row
+                break
+        (header, freq), *_ = Counter(column_headers).most_common(1)
+        if freq > 1:
+            raise ValueError(f"Duplicated column header in CSV file: {header}")
+        keys.check_differences(frozenset(column_headers))
+        with NamedTemporaryFile() as tmp:
+            with gzip.open(tmp.name, "wb") as f_out:
+                with open(path, "rb") as f_in:
+                    shutil.copyfileobj(f_in, f_out)
+                f_out.flush()
```

Comment:
Is it possible to accomplish this encoding with a buffer rather than a write to disk?